frontend/src/metabase/query_builder/components/GuiQueryEditor.jsx

```diff
@@ -187,25 +221,23 @@ export default class GuiQueryEditor extends Component {
     }
 
     renderBreakouts() {
-        const { datasetQuery: { query }, tableMetadata, features } = this.props;
+        const { query, setDatasetQuery, features } = this.props;
 
         if (!features.breakout) {
             return;
         }
 
-        const enabled = tableMetadata && tableMetadata.breakout_options.fields.length > 0;
-        const breakoutList = [];
+        const tableMetadata = query.tableMetadata();
 
-        if (enabled) {
-            const breakouts = Query.getBreakouts(query);
+        const breakoutList = [];
 
-            const usedFields = {};
-            for (const breakout of breakouts) {
-                usedFields[Query.getFieldTargetId(breakout)] = true;
-            }
+        const enabled = tableMetadata && tableMetadata.breakout_options.fields.length > 0;
+        if (enabled && tableMetadata) {
+            const breakouts = query.breakouts();
 
-            const remainingFieldOptions = Query.getFieldOptions(tableMetadata.fields, true, tableMetadata.breakout_options.validFieldsFilter, usedFields);
-            if (remainingFieldOptions.count > 0 && (breakouts.length === 0 || breakouts[breakouts.length - 1] != null)) {
+            const dimensions = query.breakoutableDimensions();
+            if (dimensions.count > 0 && (breakouts.length === 0 || breakouts[breakouts.length - 1] != null)) {
```

Comment:
This looks really fragile to me.  Is there a way to not have quite so many things to check for?